axelor-production/src/main/java/com/axelor/apps/production/service/batch/BatchFillSopRealValues.java

```diff
@@ -0,0 +1,180 @@
+import com.axelor.apps.stock.db.repo.StockLocationLineHistoryRepository;
+import com.axelor.i18n.I18n;
+import com.google.inject.Inject;
+import com.google.inject.persist.Transactional;
+import java.math.BigDecimal;
+import java.math.MathContext;
+import java.math.RoundingMode;
+
+public class BatchFillSopRealValues extends AbstractBatch {
+
+  protected SaleOrderLineRepository saleOrderLineRepo;
+  protected ManufOrderRepository manufOrderRepo;
+  protected StockLocationLineHistoryRepository stockLocationLineHistoryRepo;
+  protected PeriodService periodService;
+  protected AppBaseService appBaseService;
+
+  @Inject
+  public BatchFillSopRealValues(
+      SaleOrderLineRepository saleOrderLineRepo,
+      ManufOrderRepository manufOrderRepo,
+      StockLocationLineHistoryRepository stockLocationLineHistoryRepo,
+      PeriodService periodService,
+      AppBaseService appBaseService) {
+    this.saleOrderLineRepo = saleOrderLineRepo;
+    this.manufOrderRepo = manufOrderRepo;
+    this.stockLocationLineHistoryRepo = stockLocationLineHistoryRepo;
+    this.periodService = periodService;
+    this.appBaseService = appBaseService;
+  }
+
+  @Override
+  protected void process() {
+    ProductionBatch productionBatch = batch.getProductionBatch();
+    computeSopRealValues(productionBatch);
+  }
+
+  @Transactional
+  protected void computeSopRealValues(ProductionBatch productionBatch) {
+    Company company = productionBatch.getCompany();
+    for (Sop sop : productionBatch.getSopSet()) {
+      computeValuesPerSOP(sop, company);
+    }
+  }
+
+  protected void computeValuesPerSOP(Sop sop, Company company) {
+    ProductCategory productCategory = sop.getProductCategory();
+    for (SopLine sopLine : sop.getSopLineList()) {
+      try {
+        computeRealSalesAndSalesGap(company, productCategory, sopLine);
+        computeRealProduction(company, productCategory, sopLine);
+        computeRealStock(company, productCategory, sopLine);
+        incrementDone();
+      } catch (Exception e) {
+        incrementAnomaly();
+        TraceBackService.trace(e, ExceptionOriginRepository.SOP, batch.getId());
+      }
+    }
+  }
+
+  protected void computeRealSalesAndSalesGap(
+      Company company, ProductCategory productCategory, SopLine sopLine) {
+    Period period = sopLine.getPeriod();
+    sopLine.setSopRealSales(
+        saleOrderLineRepo
+            .all()
+            .filter(
+                "self.saleOrder.statusSelect IN (3,4) AND self.product.productCategory = :productCategory AND self.saleOrder.company = :company AND self.saleOrder.confirmationDateTime BETWEEN :startDate AND :endDate")
+            .bind("productCategory", productCategory)
+            .bind("company", company)
+            .bind("startDate", period.getFromDate())
+            .bind("endDate", period.getToDate())
+            .fetchStream()
+            .map(SaleOrderLine::getInTaxTotal)
+            .reduce(BigDecimal.ZERO, BigDecimal::add));
```

Comment:
I think there is a functionnal error here, you can check with Mathis but saleOrderLine.getInTaxTotal is not always in the same currency so doing a sum without taking into account the currency is not correct. So instead you may use companyInTaxTotal
Also please if it is not too difficult, use directly a query (createQuery method) to compute the sum instead of using a fetchStream.